core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -360,6 +361,168 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
   }
 
   @Test
+  def testCreatePartitions(): Unit = {
+    client = AdminClient.create(createConfig)
+
+    // Create topics
+    val topic1 = "create-partitions-topic-1"
+    TestUtils.createTopic(zkUtils, topic1, 1, 1, servers, new Properties)
+
+    // assert that the topic has 1 partition
+    assertEquals(1, client.describeTopics(Set(topic1).asJavaCollection).values.get(topic1).get.partitions.size)
+
+    val validateOnly = new CreatePartitionsOptions().validateOnly(true)
+
+    // assert that a validateOnly request doesn't increase the number of partitions
+    var alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(2))).asJava, validateOnly)
+    var altered = alterResult.values.get(topic1).get
+    // assert that the topics still has 1 partition
+    assertEquals(1, client.describeTopics(Set(topic1).asJavaCollection).values.get(topic1).get.partitions.size)
+
+    // try creating a new partition (no assignments), to bring the total to 2 partitions
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(2))).asJava)
+    altered = alterResult.values.get(topic1).get
+    // assert that the topics now has 2 partitions
+    var actualPartitions = client.describeTopics(Set(topic1).asJavaCollection).values.get(topic1).get.partitions
+    assertEquals(2, actualPartitions.size)
+
+    // now try creating a new partition (with assignments), to bring the total to 3 partitions
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(3), asList(asList(new Integer(0))))).asJava)
+    altered = alterResult.values.get(topic1).get
+    // assert that the topics now has 3 partitions
+    actualPartitions = client.describeTopics(Set(topic1).asJavaCollection).values.get(topic1).get.partitions
+    assertEquals(3, actualPartitions.size)
+    assertEquals(List(new Integer(0)), actualPartitions.get(2).replicas.asScala.map(_.id).toList)
+
+    // try a newCount which would be a decrease
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(1))).asJava, validateOnly)
+    try {
+      alterResult.values.get(topic1).get
+      fail("Expect InvalidPartitionsException when newCount is a decrease")
+    } catch {
+      case e: ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[InvalidPartitionsException])
+        assertEquals("Topic currently has 3 partitions: Cannot create -2 partitions.", e.getCause.getMessage)
+    }
+
+    // try a newCount which would be a noop
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(3))).asJava, validateOnly)
+    try {
+      alterResult.values.get(topic1).get
+      fail("Expect InvalidPartitionsException when newCount == oldCount")
+    } catch {
+      case e: ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[InvalidPartitionsException])
```

Comment:
How do we detect a no-op? If we provide a way to do that, then one doesn't have to `describe` to know if the operation succeeded. Another option is that a no-op doesn't return an error. Pros and cons to that, so not sure.

Reply:
The API is to create partitions, so its simply an error if the increment is 0 (or negative). See `AdminManager.createPartitions()`

Reply:
I think it's important to think about how users will use the API. A common case is that someone wants to retry until the number of partitions is 10. Let's assume that the first call times out, but actually succeeds. On the second call, the user gets an error even though the desired result was achieved. How do you envisage the user dealing with this? Do they have to `describeTopics` to verify? It would be convenient to be able to handle this case without the `describe`, I think.

Reply:
Btw, what I'm suggesting is quite common for idempotent APIs.

Reply:
Well that would be OK in the case that no replica assignment is given, but if replica assignments are given we ought to return an error (because otherwise the user will get the impression that those assignments are in effect when in fact they might not be). But then you have an inconsistency in the API between the cases where assignments are given or not.